README


This archive contains the distribution MooX-Role-CachedURL,
version 0.06:

  a role providing a locally cached copy of a remote file

This software is copyright (c) 2014 by Neil Bowers.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.015.



Changes
Revision history for Perl module MooX::Role::CachedURL

0.06 2014-05-05
    - Changed tests 2 and 3 to use a cache path in t/.
      Two of the tests were reported as failing on some CPAN Testers
      platforms (inc Linux), when unlinking cache files just created.
      Updated the code doing the unlinking to report why it fails,
      so if the tests still fail, I'll get an idea why.

0.05 2014-05-05
    - Added a test for the 'path' attribute: t/04-path.t
    - If the cache_path or path attributes reference a file ending with .gz,
      we assume it's gzip'd and transparently unzip it for you
      with PerlIO::gzip. This means that large files can be requested and
      cached gzip'd.

0.04 2014-04-28
    - Split the 'path' into 'path' and 'cache_path' attributes.
      Previously 'path' was used for both the local cache path,
      and for your to specify your own local file. That meant you
      couldn't specify your own path for local caching.
    - Added [Git::Tag] and [Git::Push] to dist.ini

0.03 2014-03-29
    - Changed the 'expires' attribute to 'max_age', which feels
      the right name having read various caching-related things recently.

0.02 2014-03-25
    - Added 'expires' attribute, which uses Time::Duration::Parse to
      let you express the expiry time in a variety of ways.
    - Split out test package into a separate module in t/lib/
    - Added major todo items to section in pod

0.01 2014-03-22
    - First release to CPAN



t/lib/CPAN/Robots.pm
package
    CPAN::Robots;

use Moo;
with 'MooX::Role::CachedURL';

has '+url' => (default => sub { 'http://www.cpan.org/robots.txt' });

1;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "a role providing a locally cached copy of a remote file",
  "AUTHOR" => "Neil Bowers <neil\@bowers.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "MooX-Role-CachedURL",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MooX::Role::CachedURL",
  "PREREQ_PM" => {
    "Carp" => 0,
    "File::HomeDir" => 0,
    "File::Spec::Functions" => 0,
    "HTTP::Tiny" => 0,
    "Moo::Role" => 0,
    "PerlIO::gzip" => 0,
    "Time::Duration::Parse" => 0
  },
  "TEST_REQUIRES" => {
    "File::Touch" => 0,
    "Moo" => 0,
    "Test::File" => 0,
    "Test::More" => "0.88",
    "autodie" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.06",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "File::HomeDir" => 0,
  "File::Spec::Functions" => 0,
  "File::Touch" => 0,
  "HTTP::Tiny" => 0,
  "Moo" => 0,
  "Moo::Role" => 0,
  "PerlIO::gzip" => 0,
  "Test::File" => 0,
  "Test::More" => "0.88",
  "Time::Duration::Parse" => 0,
  "autodie" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/MooX/Role/CachedURL.pm
package MooX::Role::CachedURL;
$MooX::Role::CachedURL::VERSION = '0.06';
use 5.006;
use Moo::Role;
use File::HomeDir;
use File::Spec::Functions 'catfile';
use HTTP::Tiny;
use Time::Duration::Parse qw/ parse_duration /;
use Carp;

has 'url'           => (is => 'ro');
has 'path'          => (is => 'ro');
has 'cache_path'    => (is => 'rw');
has 'max_age'       => (is => 'ro');

sub BUILD
{
    my $self = shift;

    if ($self->path) {
        return if -f $self->path;
        croak "the file you specified with 'path' doesn't exist";
    }

    # If constructor didn't specify a local file, then mirror the file
    if (not $self->cache_path) {
        my $basename = $self->url;
           $basename =~ s!^.*[/\\]!!;

        my $classid  = ref($self);
           $classid  =~ s/::/-/g;

        $self->cache_path( catfile(File::HomeDir->my_dist_data( $classid, { create => 1 } ), $basename) );
    }

    if (-f $self->cache_path && defined($self->max_age)) {
        my $max_age_in_seconds = parse_duration($self->max_age);
        return unless time() - $max_age_in_seconds > (stat($self->cache_path))[9];
    }

    my $response;
    eval { $response = HTTP::Tiny->new()->mirror($self->url, $self->cache_path) };
    if (not $response->{success}) {
        croak "failed to mirror @{[ $self->url ]}: $response->{status} $response->{reason}";
    }

}

sub open_file
{
    my $self     = shift;

    my $layers   = ':encoding(UTF-8)';
    my $filename = defined($self->path)
                   ? $self->path
                   : $self->cache_path;

    if ($filename =~ /\.gz\z/) {
        require PerlIO::gzip;
        $layers = ':gzip'.$layers;
    }

    open(my $fh, '<'.$layers, $filename)
        || croak "can't open $filename: $!";

    return $fh;
}

sub close_file
{
    my $self = shift;
    my $fh   = shift;

    close($fh);
}

1;

=head1 NAME

MooX::Role::CachedURL - a role providing a locally cached copy of a remote file

=head1 SYNOPSIS

 package MyClass;
 use Moo;
 with 'MooX::Role::CachedURL';
 has '+url' => (default => sub { 'http://www.cpan.org/robots.txt' });
 
 sub my_method {
    my $self = shift;
    my $fh   = $self->open_file;
 
    while (<$fh>) {
        ...
    }
    $self->close_file($fh);
 }

Then in the user of MyClass:

 use MyClass;
 my $object = MyClass->new(max_age => '2 days');
  
 print "local file is ", $object->cache_path, "\n";

=head1 DESCRIPTION

This role represents a remote file that you want to cache locally,
and then process.
This is common functionality that I'm pulling out of my L<PAUSE::Users>,
L<PAUSE::Permissions> and L<PAUSE::Packages> modules.

PAUSE::Users provides a simple interface to the C<00whois.xml> file
that is generated by PAUSE.
It caches the file locally,
then provides a mechanism for iterating over all users in the file.

=head1 ATTRIBUTES

=head2 cache_path

The full path to the local file where the content of the remote URL
will be cached. You can provide your own path, but if you don't,
then an appropriate path for your operating system will be generated.

=head2 path

A full or relative path to your own copy of the cached content.
If you provide this, then your content will be used,
without checking the remote URL.
If the file you pass doesn't exist, the module will C<croak()>.

=head2 url

This specifies the URL that should be cached locally.
It should be over-ridden in the composing class, as shown in the SYNOPSIS above.

=head2 max_age

Specifies the maximum age of the local copy, in seconds.
We won't even look for a new remote copy if the cached copy is younger than this.

You can specify max_age using any of the notations supported by L<Time::Duration::Parse>.
For example:

 max_age => '2 hours',

=head1 Support for gzip'd files

If the C<cache_path> or C<path> attribute ends in C<.gz>,
then the file is assumed to be gzip'd, and will be transparently handled
using L<PerlIO::gzip>.

=head1 TODO

=over 4

=item * Switch to LWP for general URL handling, not just HTTP

=item * Ability for a class to transform content when caching

=back

=head1 REPOSITORY

L<https://github.com/neilbowers/MooX-Role-CachedURL>

=head1 AUTHOR

Neil Bowers E<lt>neilb@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Neil Bowers <neilb@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
